docs/_ext/doc_redirect.py

```diff
@@ -0,0 +1,169 @@
+from sphinx.util import logging
+
+LOGGER = logging.getLogger(__name__)
+
+TEMPLATE_REDIRECT = """<html>
+    <head>
+        <noscript>
+            <meta http-equiv="refresh" content="0; url={to_path}"/>
+        </noscript>    
+    </head>
+    <body>
+        <script>
+            window.location.href = '{to_path}' + (window.location.search || '') + (window.location.hash || '');
+        </script>
+        <p>You should have been redirected.</p>
+        <a href="{to_path}">If not, click here to continue.</a>
+    </body>
+</html>
+"""
+TEMPLATE_CODE = """<html>
+    <body>
+        <pre>
+{code}
+        </pre>
+    </body>
+</html>
+"""
+
+
+def generate_redirects(app):
+    """
+    Generates HTML page redirections between different file path locations.
+
+    This extension allows to fake a link to the HTML page actually generated by an RST file which contains a literal
+    *reference* to another RST file.
+
+    For example, suppose we have two RST (ie: ``file1.rst`` and ``file2.rst`` ). Within ``file1.rst`` we got:
+
+    .. code-block:: rst
+
+        see `file2`_ details
+
+        .. _file2: ./docs/file2.rst
+
+    Normally, the generated HTML will have an hyperlink named ``file2`` with a *literal* reference to ``file2.rst``.
+    This will result in HTTP Not Found (404) as it doesn't correspond to the generated ``file2.html``. Normally, this
+    can be fixed using the following directive:
+
+    .. code-block:: rst
+
+        :doc:`./docs/file2.rst`
+
+    But then, rendering on GitHub becomes literally this string with an hyperlink reference that doesn't lead to the
+    desired ``file2.rst`` (for quick documentation locally within the GitHub repository).
+
+    With this extension, if configuration is specified as follows, the HTML link is resolved by redirecting the literal
+    ``./file2.rst`` reference in the output HTML build directory to the desired ``file2.html`` generated (as required).
+
+    .. code-block:: python
+
+        doc_redirect_map = {
+            # mapping of: <actual-reference-in-rst> => <pointed-rst-generated-as-html>
+            "file2.rst": "file2.rst"
+        }
+
+    In other words, when ``file1.rst`` is viewed from GitHub, the literal relative path is used an the file is found,
+    while on ``Readthedocs`` (or when viewing locally in a browser), ``file1.html`` will contain a raw relative path to
+    where the pointed ``file2.html`` *should* be using corresponding base directories. This is demonstrated below:
+
+    .. code-block:: text
+
+        '<root>/docs/file1.rst' ===> '<build-html>/file1.html' (ref [file2]) ---> (raw '<root>/docs/file2.rst')
+                                                                                               |
+                                                                                               |
+        '<root>/docs/file2.rst' ===> '<build-html>/file2.html'   <------------------------------
```

Comment:
Very interesting !

Reply:
The only problem I have add so far is to try to make it work correctly under ReadTheDocs. 
Locally, it redirects fine, but ReadTheDocs triggers a file download of the "RST" with HTML contents. 
Still exploring...

Reply:
You might want to ask the ReadTheDocs developper.  They might hardcode some behavior on ReadTheDocs and let you know how to undo it.  They might even suggest a different solution to our problem that could be more compatible.

If they have a compatible solution that allow us to avoid those ugly looking duplicate download link, this would be huge.

Reply:
Good idea.

Reply:
Maybe an approach would simply to define redirects specifically in ReadTheDocs: 
https://docs.readthedocs.io/en/latest/user-defined-redirects.html
It's somewhat manual configuration though. 

Others have tried to automate it, but it's not trivial (cannot define it via config yet).
https://github.com/readthedocs/readthedocs.org/issues/2904